Restart

RadomirVavra opened the pull request:
Implementoval jsem zatím base a input třídy, ještě se pustím do output. Vyšel jsem z poslední verze, ale musel jsem to hodně upravit. Rozhodl jsem se to zjednodušit tím, že jsem odstranil ten asi těžko pochopitelný a vysvětlitelný princip s tiles. Tím jsou data pouze 4D, kódy o moc jednodušší nejsou, ale může to zvýšit aplikační potenciál, když tomu budou lidé lépe rozumět. Tiles bych doimplementoval pomocí dědičnosti. Zároveň se tím ověří, že je vše vymyšleno co nejlépe a bude možné dědičností přidávat i další dimenze. Připravím také ještě TODO list. Bude tam zejména, že je potřeba napsat testovací metody na ty core třídy, implementovat čtení ze souboru po jednotlivých entitách, pokud je málo místa v paměti a implementovat pomocí dědičnosti BigBase, BigInput a BigOutput, které už umí pracovat se všemi meta-daty z dokumentace. Libore, pokud budeš chtít, můžeš se do něčeho z toho rovnou pustit. Zejména to testování bych teď ocenil.

libka-b committed:
Fixed and commented some issues.

libka-b committed:
uint8_t -> uint64_t. Deleted resolved comments.

libka-b committed:
Fixed to work the way we agreed to. Missing methods to implement: getTile.

libka-b committed:
Solution for dataOrder = (1, 2, *, *, *)

RadomirVavra committed:
first changes

RadomirVavra committed:
Specification of the BIG format simplified - concept of tiles removed.
Documentation modified appropriately.
Base and Input classes implemented.
Tiles will be implemented through inheritance.

libka-b commented on src/big_core_base.cpp:
```diff
@@ -1,120 +1,99 @@
+            dimensions.push_back(numberOfPlanes);
+            break;
+        }
+    }
```
Pri pouziti iteratoru bych radsi psal `for(auto it : dataOrder)` je to citelnejsi a preklada se to do totozneho byte kodu.

RadomirVavra replied:
Máš pravdu, upravil jsem to, díky.

libka-b commented on src/big_core_input.cpp:
```diff
@@ -5,316 +5,217 @@
+    if (numberOfImages == 0) throw "Number of images in the file is zero!";
+    if (imageHeight == 0) throw "Height of images in the file is zero!";
+    if (imageWidth == 0) throw "Width of images in the file is zero!";
+    if (dataPosition == 0) throw "Data are missing in the file!";
```
Tohle vzdycky vyhodi `"Data are missing in the file!"`. Nikde neupdatujes dataPosition

RadomirVavra replied:
Na řádku 39 (if (!readChunk(file, id, length) || !readData(file, id, length)) {) se čtou všechna meta-data ze souboru a dataPosition by se mělo updatovat.

libka-b replied:
No jo jasne, to je tridni promenna. Je to hrozne matouci bez toho `this->`

libka-b commented on include/big_core_base.h:
```diff
@@ -45,78 +54,105 @@ enum class DataTypes : uint8_t // RV: v big.pdf je jasne definovane, ze je to 8
+    char *_data = nullptr;
+
+    uint64_t dataSize = 0;                              // size of the data according to given dimensions and data types
+    uint64_t memorySize = 0;                            // size of the used memory
```
k cemu je memorySize, kdyz ve chvili kdy se data do pameti nevejdou, tak je vubec nenacitas?

RadomirVavra replied:
V budoucnu bych chtěl mít alokovaný buffer (o velikosti memorySize), do kterého se budou data číst v případě, že je po nich poptávka. Do bufferu se budou ukládat jednotlivé outermost entity. Pokud bude dataOrder = {1,*,*,*}, budou se do bufferu načítat vždy celé obrázky. Pokud například chci pixel z obrázku číslo 25, načte se celý do bufferu a vrátí se daný pixel. Pokud pak v budoucnu budu chtít zase nějaký pixel z obrázku 25, sáhne se do bufferu. Když v bufferu dojde místo, nahradí se nejdelší dobu nepoužitý obrázek. Pokud uživatel nechce buffer používat, nastaví si maxMemorySize = 0. Trochu problém bude, že různé obrázky můžou zabírat různou velikost, protože můžou mít různý datový typ. Pak bude nahrání obrázku s typem double do plného bufferu kde jsou samé obrázky typu char trochu problém, protože se bude muset zahodit 8 obrázků vedle sebe, aby se tam ten nový vešel. To si musím ještě rozmyslet. Možná by bylo nejlepší alokovat pro každý obrázek novou nezávislou paměť. Pak když už velikost alokované paměti dosáhne maxima, můžu jednoduše uvolnit 8 obrázků typu char a alokovat jeden typu double. Bude to o něco pomalejší, protože se občas bude muset alokovat nová paměť. Ale zase to nebude muset na začátku alokovat všechnu paměť, pouze to bude postupně přidávat novou paměť.

libka-b commented on src/big_core_base.cpp:
```diff
@@ -1,120 +1,99 @@
-bool BigCoreBase::isEmpty()
-{
-	return this->dataLength == 0 ? true : false;
+    delete[] _data;
```
Takhle nemuzes. Pokud _data nejsou nactena, tak se snazis uvolnovat nealokovanou pamet. Spravne je
```
if(this->_data)
    delete [] this->_data;
```

RadomirVavra replied:
Podle specifikace to může být nullptr: If this is a null-pointer, the function does nothing. Podmínka je tedy zbytečná.

libka-b replied:
Podle jake specifikace?
```
BigCoreBase::~BigCoreBase()
{
    delete[] _data;
}
```
Tohle evidentne neco dela. Destruktor se vola pri zahazovani lokalni promenne v ramci scopu. Pokud tedy pamet nebude nactena a bude se rusit instance tridy BigCore, bude se volat destruktor a program skonci segmentation faultem.

RadomirVavra replied:
Pokud _data == nullptr (což se nastaví na začátku libovolného konstruktoru třídy BigCoreBase), pak delete[] _data neudělá nic, tedy žádnou chybu to nezpůsobí. Pokud _data != nullptr, pak delete[] _data uvolní paměť. Podmínka navíc by nic nového nepřidala. Specifikace například zde: http://www.cplusplus.com/reference/new/operator%20delete[]/

libka-b replied:
priklad: vytvori se instance tridy, vse se nacte, neco se s tim udela. Budu chtit prepouzit tuhle instanci, predam novy soubor, data se nebudou vejit do pameti, _data budou uvolnena. Pri volani destruktoru segmentation fault.

Viz:
test.cpp:
```
int main()
{
    int* arr = new int[5];
    delete[] arr;
    delete[] arr;
    return 0;
}
```
`g++ -o test test.cpp -Wall -std=c++11 -pedantic -g`
./test
*** Error in `./test': double free or corruption (fasttop): 0x0000000001966010 ***
Segmentation fault (SIGABRT) (core dumped)

RadomirVavra replied:
Ano, tvůj test.cpp samozřejmě způsobí segmentation fault. Ale jak by se to mohlo stát s třídou Big? Pokud předám nový soubor, půjde to přes metodu openFile(fileName) (jediné další kam jde předat ten soubor, jsou metody readFile(fileName) a readFile(fileName, bytes), ale ty obě volají metodu openFile). Metoda openFile pokud soubor otevře a bude se shodovat hlavička, zavolá metodu clear(), které provede delete[] _data a nastaví _data = nullptr. Pak když se zavolá destruktor, tak to projde korektně, protože _data == nullptr. Pokud vidíš ještě jiný scénář, kdy by to mohlo skončit chybou, popiš to prosím podrobně, protože já v tom nikde problém nemůžu najít.

libka-b replied:
mas pravdu, toho nastavovani _data = nullptr v clear() jsem si nevsim

libka-b commented on src/big_core_base.cpp:
```diff
@@ -1,120 +1,99 @@
+    numberOfPlanes = 1;
+    dataOrder = { DataOrderIds::NUMBER_OF_IMAGES, DataOrderIds::IMAGE_HEIGHT, DataOrderIds::IMAGE_WIDTH, DataOrderIds::NUMBER_OF_PLANES };
+    dataType = { DataTypes::FLOAT };
+    delete[] _data;
```
stejny pripad jako u toho destruktoru

RadomirVavra replied:
Podle specifikace to může být nullptr: If this is a null-pointer, the function does nothing. Podmínka je tedy zbytečná.

RadomirVavra committed:
Implemented BigCoreOutput class.
Fixed minor issues.

RadomirVavra commented:
Implementoval jsem output třídu. Samozřejmě to stále neumí pracovat s tím, že se to nevejde do paměti. Na tom teď asi budu pracovat, aby Core třídy byly kompletní. Pak bude možné implementovat dědící třídy pro ostatní meta-data a dále rozšíření s tiles.

libka-b commented:
Ja se k tomu tenhle tyden dostanu jenom kdyz budu mit obcas v praci chvilku, takze muzu delat maximalne code review.

RadomirVavra committed:
Implemented BigCache. Added ToDo list.

RadomirVavra commented:
Díky alespoň za to code review. Implementoval jsem zatím nějak tu cache, pak to bude chtít trochu optimalizovat. Kvůli cache se vše docela komplikuje. V todo listu jsem popsal, jak to vlastně vytváří tři módy jak přistupovat k datům. Momentálně bude nejdůležitější udělat testovací metody, aby se to alespoň pro ten případ, kdy je všechno v paměti, mohlo začít používat. Ale ještě asi upravím tu filozofii ukládání dat, že každá entita bude mít vlastní chunk. Tím to pak umožní přidávat a mazat entity a na výkon by to nemuselo mít vliv. Dej prosím vědět, zda budeš moct udělat nějaké ty testovací metody. Já se k tomu dostanu zase až příští týden.

libka-b commented on include/big_core_output.h:
```diff
@@ -4,125 +4,251 @@
+
+    uint64_t index = offsets[indices[0]] + (indices[1] * subSize[1] + indices[2] * subSize[2] + indices[3]) * entityTypeSizes[indices[0]];
+    return static_cast<T&>(_data[index]);
+}
```
tyhle metody jsou stejne jako v BigCoreInput, dal bych je do BigCoreBase a tady se prida jenom override na 
```c++
changed = true;
BigCoreBase::operator()(imageNum, row, col, plane);
```

libka-b commented on src/big_core_output.cpp:
```diff
@@ -1 +1,184 @@
+    if (dataPosition == 0) {
+        const auto id = CoreChunkIds::DATA;
+        file.write(reinterpret_cast<const char*>(&id), CHUNK_LENGTH);
+        const uint64_t n = static_cast<uint64_t>(std::ceil(dataSize / static_cast<double>(CHUNK_LENGTH)));
```
proc tady delas `static_cast<double>(CHUNK_LENGTH)`?

libka-b commented on src/big_core_output.cpp:
```diff
@@ -1 +1,184 @@
+        file.write(_data, dataSize);
+        const uint64_t zero = 0;
+        file.write(reinterpret_cast<const char*>(&zero), CHUNK_LENGTH - dataSize % CHUNK_LENGTH);
+    }
```
tady bys mel nastavit changed = false, nebo locked = true, jinak kdokoliv muze zavolat reserve() a data se zapisou znova

RadomirVavra committed:
Re-implemented once more.
BigCoreInput and BigCoreOutput have now more meaningfull interface.
BigCoreInput returns direct value (operator(), method at), fills data array (methods getData and getImage), returns direct (but const) shared pointer to entity in cache (operator[], method at).
BigCoreOutput takes direct value (methods addElement, addElementFast), takes data array (method addData, addImage), takes shared pointer to entity - this is potentionally dangerous, but fast.
Constants moved to big_core_consts.h. We should probably use namespace or something.
